### PR TITLE
Issue/2228 Vocabulary Suggestion box for keywords & themes

### LINKED ISSUE #2228: Vocabulary Tool for Keywords & Themes
As a Magda User
I want keywords to be constrained by a defined vocabulary
So that the data quality is better

### Acceptance Criteria:
1. Make keywords and themes work like predictive search /suggestions so that the user gets an idea of what is already existing in the vocabulary E.g. asic could auto-complete to ASIC. Department of Agriculture and Water Resources is always called that in that format.

### PR DESCRIPTION
**Pull request by t83714**
### What this PR does

Fixes #2228 

Changes:
- Vocabulary Suggestion box for keywords & themes
- API endpoints configurable through helm. e.g.
```yaml
web-server:
  fallbackUrl: "https://data.gov.au"
  featureFlags:
    cataloguing: true
  vocabularyApiEndpoints: 
    - "https://vocabs.ands.org.au/repository/api/lda/abares/australian-land-use-and-management-classification/version-8/concept.json"
    - "https://vocabs.ands.org.au/repository/api/lda/neii/australian-landscape-water-balance/version-1/concept.json"
    - "https://vocabs.ands.org.au/repository/api/lda/ands-nc/controlled-vocabulary-for-resource-type-genres/version-1-1/concept.json"
```
- Keyword extraction result will be filtered out according to Vocabulary API

### Checklist

-   [x]  unit tests aren't applicable
-   [x] I've updated CHANGES.md with what I changed.
-   [x] I've linked this PR to an issue in ZenHub (core dev team only)

### CODE REVIEW

File: magda-web-client/src/Components/Editing/Editors/textEditor.tsx [hunk 1]
```diff
@@ -1,9 +1,19 @@
-import React from "react";
+import React, { ReactEventHandler } from "react";
 import Editor from "./Editor";
 
 import { ListMultiItemEditor } from "./multiItem";
 
-export function textEditorEx(options: any = {}) {
+export type InputComponentParmeter = React.ComponentType<{
```
Comment (AlexGilleran) [inline magda-web-client/src/Components/Editing/Editors/textEditor.tsx]: Parameter

File: magda-web-client/src/helpers/VocabularyApis.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    const keywords: string[] = [];
+
+    data.result.items.forEach(item => {
+        let prefLabels: {
+            _value: string;
+            _lang: string;
+        }[] = [];
+
+        if (item.broader && item.broader.prefLabel) {
+            if (!Array.isArray(item.broader.prefLabel)) {
```
Comment (AlexGilleran) [inline magda-web-client/src/helpers/VocabularyApis.ts]: I think it'd work a bit better if we included the other labels too, not just the pref labels?

File: magda-web-client/src/Components/Dataset/MetadataExtraction/extractKeywords.js [hunk 3]
```diff
@@ -1,15 +1,24 @@
...
 import toString from "nlcst-to-string";
+import { isValidKeyword } from "../../../helpers/VocabularyApis";
 
 /**
  * Extract keywords from text based file formats
  */
 export async function extractKeywords(input, output) {
     if (input.text) {
-        output.keywords = (output.keywords || []).concat(
+        const candicateKeywords = (output.keywords || []).concat(
```
[the remark this answers is not in the document]
Comment (t83714) [reply magda-web-client/src/Components/Dataset/MetadataExtraction/extractKeywords.js]: 👍 fixed

File: magda-web-client/src/Components/Editing/Editors/textEditor.tsx (hunk 1 above)
Comment (t83714) [reply magda-web-client/src/Components/Editing/Editors/textEditor.tsx]: 👍 fixed

File: magda-web-client/src/helpers/VocabularyApis.ts (hunk 2 above)
Comment (t83714) [reply magda-web-client/src/helpers/VocabularyApis.ts]: As discussed, we leave this unchanged until we found other fields that want to be included

**Comment by t83714** [comment]
@AlexGilleran 
It's ready to have a look again~

**Pull request merged**